cmake/generic.cmake

```diff
@@ -331,3 +331,52 @@ function(go_test TARGET_NAME)
   add_custom_target(${TARGET_NAME} ALL DEPENDS ${TARGET_NAME}_timestamp ${go_test_DEPS})
   add_test(${TARGET_NAME} ${CMAKE_CURRENT_BINARY_DIR}/${TARGET_NAME})
 endfunction(go_test)
+
+function(pb_cc_library TARGET_NAME)
```

Comment:
Thanks for this PR! I do NOT think we need rules that compile .proto files into a specific programming language; instead, we need just a `proto_library`, similar to Bazel's https://bazel.build/blog/2017/02/27/protocol-buffers.html, to 

1. convert .proto into .h and .cc using the standard function [`protobuf_generate_cpp`](https://cmake.org/cmake/help/v3.0/module/FindProtobuf.html), and
1. build a static library from the generated C++ source files by calling `cc_library`.

I have a draft of `proto_library` at
https://github.com/PaddlePaddle/Paddle/pull/2569/files#diff-7a3600c73304487a6d5f6bd3fa3fd662R175:

```cmake
 +function(proto_library TARGET_NAME)
 +  if (NOT ${PROTOBUF_FOUND})
 +    error("Using proto_library but CMake cannot find installed protobuf.")
 +  endif()
 +  
 +  set(options STATIC static SHARED shared)
 +  set(oneValueArgs "")
 +  set(multiValueArgs SRCS DEPS)
 +  cmake_parse_arguments(proto_library "${options}" "${oneValueArgs}" "${multiValueArgs}" ${ARGN})
 +
 +  protobuf_generate_cpp(${TARGET_NAME}_PROTO_SRCS ${TARGET_NAME}_PROTO_HDRS ${proto_library_SRCS})
 +  cc_library(${TARGET_NAME} SRCS ${${TARGET_NAME}_PROTO_SRCS} DEPS ${${proto_library}_DEPS} protobuf)
 +endfunction(proto_library)
```